ui/button.js

```diff
@@ -30,389 +30,317 @@
-				ancestor = ancestor.length ? ancestor.siblings() : this.element.siblings();
-				this.buttonElement = ancestor.filter( labelSelector );
-				if ( !this.buttonElement.length ) {
-					this.buttonElement = ancestor.find( labelSelector );
-				}
-			}
-			this.element.addClass( "ui-helper-hidden-accessible" );
+	_updateIcon: function( icon ) {
+		var prepend = ( this.options.iconPosition === "top" || this.options.iconPosition === "beginning" ),
+			displayBlock = ( this.options.iconPosition === "top" ||
+				this.options.iconPosition === "bottom" );
+		if ( !this.icon ) {
+			this.icon = $( "<span>" );
+			this.iconSpace = $( "<span> </span>" );
+
+			this._addClass( this.icon, "ui-button-icon", " ui-icon" );
 
-			checked = this.element.is( ":checked" );
-			if ( checked ) {
-				this.buttonElement.addClass( "ui-state-active" );
+			if ( !this.options.showLabel ) {
+				this._addClass( "ui-button-icon-only" );
+			} else if ( displayBlock ) {
+				this._addClass( this.icon, null, "ui-icon-display-block" );
 			}
-			this.buttonElement.prop( "aria-pressed", checked );
 		} else {
-			this.buttonElement = this.element;
+			this._removeClass( this.icon, null, this.options.icon );
 		}
-	},
-
-	widget: function() {
-		return this.buttonElement;
+		this._addClass( this.icon, null, icon );
+		this.icon[ prepend ? "prependTo" : "appendTo" ]( this.element );
+		if ( !displayBlock ) {
+			this.icon[ prepend ? "after" : "before" ]( this.iconSpace );
+		}
+		return this;
 	},
 
 	_destroy: function() {
-		this.element
-			.removeClass( "ui-helper-hidden-accessible" );
-		this.buttonElement
-			.removeClass( baseClasses + " ui-state-active " + typeClasses )
-			.removeAttr( "role aria-pressed" )
-			.html( this.buttonElement.find(".ui-button-text").html() );
+		this.element.removeAttr( "role" );
 
+		if ( this.icon ) {
+			this.icon.remove();
+			this.iconSpace.remove();
+		}
 		if ( !this.hasTitle ) {
-			this.buttonElement.removeAttr( "title" );
+			this.element.removeAttr( "title" );
+		}
+		buttonCount--;
+		if ( buttonCount === 0 ) {
+			this.formElement.off( "reset." + this.widgetFullName, formResetHandler );
 		}
 	},
 
 	_setOption: function( key, value ) {
+		var iconGroup,
+			prepend = ( this.options.iconPosition === "top" || this.options.iconPosition === "beginning" );
+
+		if ( key === "icon" ) {
+			if ( value !== null ) {
+				this._updateIcon( value );
+			} else {
+				this.icon.remove();
+				this.iconSpace.remove();
+				this._removeClass( "ui-button-icon", " ui-icon-" + this.options.iconPosition );
+			}
+		}
+
+		// Make sure we can't end up with a button that has no text nor icon
+		if ( key === "showLabel" ) {
+			if ( ( !value && this.options.icon ) || value ) {
+				this._toggleClass( this._classes( "ui-button-icon-only" ), null, !value );
+				this._updateTooltip();
+			} else {
+				value = true;
+			}
+		}
+		if ( key === "iconPosition" && this.options.icon ) {
+			this._addClass( null, value );
```

Comment:
this logic should not be here at all there is no such thing as an iconPosition class any more and even if there was it would not just add the value.